### PR TITLE
explanation/faq: Expand answer about conflicts

### LINKED ISSUE #14: Clarify how leadership conflicts are resolved from the user's perspective
From https://chat.canonical.com/canonical/pl/snzmd1grwtrhjgwkseg3muyizy

@dimaqq

I'm reading dqlite docs and these two passages seem to be in conflict:
> How does Dqlite behave during conflict situations?
> Does Raft select a winning WAL to write and any others in-flight writes are aborted?
>
> There can’t be a conflict situation. Raft’s model is that only the leader can append new log entries, which translated to Dqlite means that only the leader can write new WAL frames. So this means that any attempt to perform a write transaction on a non-leader node will fail with an ErrNotLeader error (and in this case clients are supposed to retry against whoever is the new leader).

vs.
 
> However, stale reads are rare in practice. They can happen only in case a read transaction is started on a leader that was disconnected from a majority of voting nodes but that didn’t yet realise it and didn’t yet step down from leader. If in the meantime another leader was elected by the rest of the voting nodes and that leader committed a write transaction before the old leader ...

Either I'm misunderstanding stuff, or the earlier may need a slight re-write?

1. The way I see to reconcile the two is that the leader always acquires quorum at the start of a write tx (or tx upgrade). If that's the case, perhaps this needs to be stated somewhere, because that's a performance implication
2. Or if the current leader is optimistic, then it would accept a write tx, but fail only when trying to reach consensus while "sending out it's new log block to peers"
Show less

@cole-miller

Yeah, the wording is confusing. The leader tries to get a quorum for each (write) transaction, but after it appends the transaction to its own log. Then if it notices during the replication that it's not the leader anymore (i.e. sees a higher Raft term number) it aborts that transaction with a different error ("leadership lost").

The "am I the leader?" check before executing+appending a transaction is also a thing, but that's fast and doesn't involve talking to the other nodes.

@dimaqq

Right, so in that case, the old leader drops own entries form the log or somehow the old leader's log ends up with the content from the consensus.

I guess what prompted me to ask about this was that "There can’t be a conflict situation." is a very strong statement, and goes against general distributed systems learning... maybe it's better to explain at what level the conflict is resolved

@cole-miller

Agree, that first doc extract needs rewriting---thank you for catching the inconsistency!

### PR DESCRIPTION
**Pull request by cole-miller**
Closes #14. @dimaqq, how does this read for you?

Signed-off-by: Cole Miller <cole.miller@canonical.com>

### CODE REVIEW

File: explanation/faq.md [hunk 1]
```diff
@@ -10,11 +10,19 @@ Not at the moment. But [Canonical](https://www.canonical.com), the company who
 
 The v1 series will be maintained, improved and bug-fixed for the foreseeable future and backward compatibility is guaranteed.
 
-## How does Dqlite behave during conflict situations?
+## How does Dqlite behave during conflict situations? Does Raft select a winning WAL to write and any others in-flight writes are aborted?
```
Comment (dimaqq) [inline explanation/faq.md:13]: I think the original, short heading is better.

If the clarification is needed, then it belongs to one of the paragraphs below.

File: explanation/faq.md [hunk 2]
```diff
@@ -10,11 +10,19 @@ Not at the moment. But [Canonical](https://www.canonical.com), the company who
...
 
-Does Raft select a winning WAL to write and any others in-flight writes are aborted?
+Dqlite uses Raft to commit write transactions in the same order across all nodes in the cluster. Only a current Raft leader is allowed to start replicating a new write transaction, so in a healthy cluster where the leader is stable, no conflicts arise. Trying to execute a transaction on a non-leader node will automatically fail.
 
-There can’t be a conflict situation. Raft’s model is that only the leader can append new log entries, which translated to Dqlite means that only the leader can write new WAL frames. So this means that any attempt to perform a write transaction on a non-leader node will fail with an `ErrNotLeader` error (and in this case clients are supposed to retry against whoever is the new leader).
+In a cluster where leadership isn't stable (perhaps because communication between the nodes is disrupted), the Raft logs of distinct nodes can become mismatched. But Raft's rules ensure that a leader that remains in power for long enough will repair these mismatches, and in the meantime no transaction can be committed unless acknowledged by a majority of the cluster.
```
Comment (dimaqq) [inline explanation/faq.md:17]: I would hoist the "majority of the cluster" requirement up top, as it helps to form a mental model.

**Review by dimaqq — APPROVED** [review]
Looks much better, thank you!

I've left a few comments which may improve readability.

File: explanation/faq.md [hunk 3]
```diff
@@ -10,11 +10,19 @@ Not at the moment. But [Canonical](https://www.canonical.com), the company who
...
+
+From the perspective of a Dqlite client, some possible outcomes when submitting a write transaction `T` to a node `N`:
+
+* The request fails immediately because `N` is not the leader.
+* The request fails after a delay because `N` lost leadership before fully replicating it. In this case, it's guaranteed that no other transaction will observe the effects of `T`.
```
[the remark this answers is not in the document]
Comment (cole-miller) [reply explanation/faq.md:22]: It's a synchronous delay, will clarify.

File: explanation/faq.md (hunk 1 above)
Comment (cole-miller) [reply explanation/faq.md:13]: This comment made me realize that the word "majority" doesn't appear in the FAQ outside this paragraph! Good idea to make it more prominent.

File: explanation/faq.md (hunk 2 above)
Comment (cole-miller) [reply explanation/faq.md:17]: This comment made me realize that the word "majority" doesn't appear in the FAQ outside this paragraph! Good idea to make it more prominent.

**Review by dimaqq — APPROVED** [review]
Looks great!

**Pull request merged**